Fix serialization and deepcopying for wrapper subclasses

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #73078

This PR addresses broken support for serialization and deepcopy on wrapper subclasses. In particular, the old implementations of `Tensor.__reduce_ex__` and `Tensor.__deepcopy__` fail by assuming the tensor type has storage, which is not the case for wrapper subclasses. To fix this:
* For serialization, `Tensor.__reduce_ex__` is expanded to return a new reconstruction function `_rebuild_wrapper_subclass` (that calls `_make_wrapper_subclass`) for cases for tensor subclasses where no storage exists.
* For deepcopying, `Tensor.__deepcopy__` is expanded to call `clone()` for cases where no storage exists.

The goal is to provide reasonable default implementations usable for cases where a custom tensor type uses `_make_wrapper_subclass()` and at most sets some additional copyable attributes. If anything more complex is needed, users are expected to write their own `__deepcopy__` / `__reduce_ex__` implementations for their type.

Note that custom implementations for the above magic methods MUST handle copying the `_is_param` attribute that may be present on the custom tensor type instance (if we go this route for custom tensors as parameters).

Differential Revision: [D34342718](https://our.internmc.facebook.com/intern/diff/D34342718)

jbschlosser committed:
Fix serialization for wrapper subclasses

[ghstack-poisoned]

jbschlosser commented on torch/_tensor.py:
```diff
@@ -46,7 +46,9 @@ def _rebuild_from_type_v2(func, new_type, args, state):
 
-    ret = func(*args).as_subclass(new_type)
+    ret = func(*args)
+    if type(ret) is not new_type:
+        ret = ret.as_subclass(new_type)
```
This pattern comes up a few times - does it make sense to subsume the `if type(x) is not subclass_type` guard into `as_subclass()` itself?

albanD replied:
This one sounds ok as in the case where the Tensor has a storage, we do use the usual Tensor serialization to save and load, then slap the subclass and its attributes onto it.
The other instances should go though. So I don't think this is necessary.

jbschlosser committed:
Update on "Fix serialization and deepcopying for wrapper subclasses"


This PR addresses broken support for serialization and deepcopy on wrapper subclasses.
* For serialization, `Tensor.__reduce_ex__` is expanded to return a reconstruction function that calls `_make_wrapper_subclass()` for all non-meta, non-sparse cases where no storage exists.
* For deepcopying, `Tensor.__deepcopy__` is expanded to call `clone()` for cases where no storage exists.

[ghstack-poisoned]

jbschlosser committed:
Update on "Fix serialization and deepcopying for wrapper subclasses"


This PR addresses broken support for serialization and deepcopy on wrapper subclasses. In particular, the old implementations of `Tensor.__reduce_ex__` and `Tensor.__deepcopy__` fail by assuming the tensor type has storage, which is not the case for wrapper subclasses. To fix this:
* For serialization, `Tensor.__reduce_ex__` is expanded to return a new reconstruction function `_rebuild_wrapper_subclass` (that calls `_make_wrapper_subclass`) for all non-meta, non-sparse cases where no storage exists.
* For deepcopying, `Tensor.__deepcopy__` is expanded to call `clone()` for cases where no storage exists.

Differential Revision: [D34342718](https://our.internmc.facebook.com/intern/diff/D34342718)

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34342718).

jbschlosser committed:
Update on "Fix serialization and deepcopying for wrapper subclasses"


This PR addresses broken support for serialization and deepcopy on wrapper subclasses. In particular, the old implementations of `Tensor.__reduce_ex__` and `Tensor.__deepcopy__` fail by assuming the tensor type has storage, which is not the case for wrapper subclasses. To fix this:
* For serialization, `Tensor.__reduce_ex__` is expanded to return a new reconstruction function `_rebuild_wrapper_subclass` (that calls `_make_wrapper_subclass`) for cases for tensor subclasses where no storage exists.
* For deepcopying, `Tensor.__deepcopy__` is expanded to call `clone()` for cases where no storage exists.

Differential Revision: [D34342718](https://our.internmc.facebook.com/intern/diff/D34342718)

[ghstack-poisoned]

albanD commented on torch/_tensor.py:
```diff
@@ -132,7 +135,8 @@ def __deepcopy__(self, memo):
                 new_tensor.grad = self.grad.__deepcopy__(memo)
```
The new_empty call should also have a check that `type(new_tensor) is type(self)` and an error that says: default implementation of deepcopy for non-wrapper subclass only works for subclass that implement new_empty and for which that function returns another instance of the same subclass. You can override `__deepcopy__` on your subclass if that is the intended behavior.

albanD commented on torch/_tensor.py:
```diff
@@ -132,7 +135,8 @@ def __deepcopy__(self, memo):
 
             if not type(self) is Tensor:
-                new_tensor = new_tensor.as_subclass(type(self))  # type: ignore[arg-type]
+                if not type(new_tensor) is type(self):
```
This should be an assert now as the code above will raise an error if not.

albanD commented on test/test_python_dispatch.py:
```diff
@@ -1,13 +1,36 @@
 import logging
 
+# Simple wrapper tensor used in testing; defined globally so that it can be pickled.
+class SerializableWrapperTensor(torch.Tensor):
```
You can use the LoggingTensor defined above for this no?

albanD commented on test/test_python_dispatch.py:
```diff
@@ -338,6 +361,46 @@ def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
+
+    def test_deepcopy_wrapper_subclass(self) -> None:
+
+        class CloneableWrapperTensor(torch.Tensor):
```
Why not use the LoggingTensor here?

albanD reviewed (commented):
I think we will want to extend this testing to cover more common subclass implementation.
But this PR seems to be a step in the right direction.

jbschlosser commented:
> I think we will want to extend this testing to cover more common subclass implementation. But this PR seems to be a step in the right direction.

True, the custom tensors as parameters PR I'm working on has much more extensive tests run across currently 5 different subclass impls.

jbschlosser committed:
Update on "Fix serialization and deepcopying for wrapper subclasses"


This PR addresses broken support for serialization and deepcopy on wrapper subclasses. In particular, the old implementations of `Tensor.__reduce_ex__` and `Tensor.__deepcopy__` fail by assuming the tensor type has storage, which is not the case for wrapper subclasses. To fix this:
* For serialization, `Tensor.__reduce_ex__` is expanded to return a new reconstruction function `_rebuild_wrapper_subclass` (that calls `_make_wrapper_subclass`) for cases for tensor subclasses where no storage exists.
* For deepcopying, `Tensor.__deepcopy__` is expanded to call `clone()` for cases where no storage exists.

The goal is to provide reasonable default implementations usable for cases where a custom tensor type uses `_make_wrapper_subclass()` and at most sets some additional copyable attributes. If anything more complex is needed, users are expected to write their own `__deepcopy__` / `__reduce_ex__` implementations for their type.

Note that custom implementations for the above magic methods MUST handle copying the `_is_param` attribute that may be present on the custom tensor type instance (if we go this route for custom tensors as parameters).

Differential Revision: [D34342718](https://our.internmc.facebook.com/intern/diff/D34342718)

[ghstack-poisoned]

jbschlosser committed:
Update on "Fix serialization and deepcopying for wrapper subclasses"


This PR addresses broken support for serialization and deepcopy on wrapper subclasses. In particular, the old implementations of `Tensor.__reduce_ex__` and `Tensor.__deepcopy__` fail by assuming the tensor type has storage, which is not the case for wrapper subclasses. To fix this:
* For serialization, `Tensor.__reduce_ex__` is expanded to return a new reconstruction function `_rebuild_wrapper_subclass` (that calls `_make_wrapper_subclass`) for cases for tensor subclasses where no storage exists.
* For deepcopying, `Tensor.__deepcopy__` is expanded to call `clone()` for cases where no storage exists.

The goal is to provide reasonable default implementations usable for cases where a custom tensor type uses `_make_wrapper_subclass()` and at most sets some additional copyable attributes. If anything more complex is needed, users are expected to write their own `__deepcopy__` / `__reduce_ex__` implementations for their type.

Note that custom implementations for the above magic methods MUST handle copying the `_is_param` attribute that may be present on the custom tensor type instance (if we go this route for custom tensors as parameters).

Differential Revision: [D34342718](https://our.internmc.facebook.com/intern/diff/D34342718)

[ghstack-poisoned]

jbschlosser committed:
Update on "Fix serialization and deepcopying for wrapper subclasses"


This PR addresses broken support for serialization and deepcopy on wrapper subclasses. In particular, the old implementations of `Tensor.__reduce_ex__` and `Tensor.__deepcopy__` fail by assuming the tensor type has storage, which is not the case for wrapper subclasses. To fix this:
* For serialization, `Tensor.__reduce_ex__` is expanded to return a new reconstruction function `_rebuild_wrapper_subclass` (that calls `_make_wrapper_subclass`) for cases for tensor subclasses where no storage exists.
* For deepcopying, `Tensor.__deepcopy__` is expanded to call `clone()` for cases where no storage exists.

The goal is to provide reasonable default implementations usable for cases where a custom tensor type uses `_make_wrapper_subclass()` and at most sets some additional copyable attributes. If anything more complex is needed, users are expected to write their own `__deepcopy__` / `__reduce_ex__` implementations for their type.

Note that custom implementations for the above magic methods MUST handle copying the `_is_param` attribute that may be present on the custom tensor type instance (if we go this route for custom tensors as parameters).

Differential Revision: [D34342718](https://our.internmc.facebook.com/intern/diff/D34342718)

[ghstack-poisoned]

jbschlosser commented on test/test_serialization.py:
```diff
@@ -869,6 +869,9 @@ def __new__(cls, elem, *args, **kwargs):
         r.elem = elem
         return r
 
+    def clone(self):
+        return type(self)(self.elem.clone())
+
```
@albanD just realized these tests were here - I guess they didn't catch the case this PR addresses because the tensors tested here are old-style wrapper tensors with meta tensors underneath. Should we get rid of these tests to avoid duplication? My preference is to keep tests for all semantics related to tensor subclasses together - wdyt? Alternatively, maybe leave them in for now and clean them up later when the more comprehensive tests land?

albanD reviewed (approved):
SGTM

cc @ezyang FYI

jbschlosser committed:
Update on "Fix serialization and deepcopying for wrapper subclasses"


This PR addresses broken support for serialization and deepcopy on wrapper subclasses. In particular, the old implementations of `Tensor.__reduce_ex__` and `Tensor.__deepcopy__` fail by assuming the tensor type has storage, which is not the case for wrapper subclasses. To fix this:
* For serialization, `Tensor.__reduce_ex__` is expanded to return a new reconstruction function `_rebuild_wrapper_subclass` (that calls `_make_wrapper_subclass`) for cases for tensor subclasses where no storage exists.
* For deepcopying, `Tensor.__deepcopy__` is expanded to call `clone()` for cases where no storage exists.

The goal is to provide reasonable default implementations usable for cases where a custom tensor type uses `_make_wrapper_subclass()` and at most sets some additional copyable attributes. If anything more complex is needed, users are expected to write their own `__deepcopy__` / `__reduce_ex__` implementations for their type.

Note that custom implementations for the above magic methods MUST handle copying the `_is_param` attribute that may be present on the custom tensor type instance (if we go this route for custom tensors as parameters).

Differential Revision: [D34342718](https://our.internmc.facebook.com/intern/diff/D34342718)

[ghstack-poisoned]

albanD reviewed (approved):
SGTM

jbschlosser commented:
@pytorchbot merge this please

jbschlosser committed:
Update on "Fix serialization and deepcopying for wrapper subclasses"


This PR addresses broken support for serialization and deepcopy on wrapper subclasses. In particular, the old implementations of `Tensor.__reduce_ex__` and `Tensor.__deepcopy__` fail by assuming the tensor type has storage, which is not the case for wrapper subclasses. To fix this:
* For serialization, `Tensor.__reduce_ex__` is expanded to return a new reconstruction function `_rebuild_wrapper_subclass` (that calls `_make_wrapper_subclass`) for cases for tensor subclasses where no storage exists.
* For deepcopying, `Tensor.__deepcopy__` is expanded to call `clone()` for cases where no storage exists.

The goal is to provide reasonable default implementations usable for cases where a custom tensor type uses `_make_wrapper_subclass()` and at most sets some additional copyable attributes. If anything more complex is needed, users are expected to write their own `__deepcopy__` / `__reduce_ex__` implementations for their type.

Note that custom implementations for the above magic methods MUST handle copying the `_is_param` attribute that may be present on the custom tensor type instance (if we go this route for custom tensors as parameters).

Differential Revision: [D34342718](https://our.internmc.facebook.com/intern/diff/D34342718)

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge this please
